### PR TITLE
Implement eventmanager and deprecate withdrawalWaitgroup

### PR DESCRIPTION
**Pull request by joe-bowman**
## 1. Summary
Fixes potential race conditions in redemption rate calculation.

## 2.Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## 3. Implementation details

Instead of relying on a single counter that is used to gate execution of logic such as redemption rate calculations, we should track each emitted message and query, and their receipt/ack. This will allow for better reasoning as to the state of the application at any given point in time and aid debugging. It also hugely reduces the complexity of nested callbacks, as all actions can be triggered ahead of time, and future actions conditional on state can be queued awaiting acknowledgements.

Events can be created for each emission; future events may be queued up with execution conditions, a callback function and a payload.

We can now, for example, trigger a redemption rate calculation to only be computed once all unbonding, redelegation, delegation, withdrawals and delegation queries have been completed.

Execution conditions may be complex and nested. Initially we support ConditionAny (any of the fields match), ConditionAll (all fields match), ConditionOr (one or more conditions must b true) and ConditionAnd (both condition A and B must resolve to true).

### CODE REVIEW

**Comment by joe-bowman** [comment]
Todo:
- ConditionAny
- Queries
- Wire up interchainstaking to use EM

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>ab4d3dc</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/6maAxXMfVd3dfnA2594F59j9NqtS'>https://vercel.com/quicksilver/quicksilver/6maAxXMfVd3dfnA2594F59j9NqtS</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8576786237)

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>c602d6f</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/x5pS41wBWKKqQcmNpgkKX5GSQyBz'>https://vercel.com/quicksilver/quicksilver/x5pS41wBWKKqQcmNpgkKX5GSQyBz</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8576793834)

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>8115147</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/8BGDVEj2FPKfDWmnUtwXTgFB6Eou'>https://vercel.com/quicksilver/quicksilver/8BGDVEj2FPKfDWmnUtwXTgFB6Eou</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8576848794)

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>5867948</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/BoxY95sB4fKQdi7Rdb7yDucyGS2y'>https://vercel.com/quicksilver/quicksilver/BoxY95sB4fKQdi7Rdb7yDucyGS2y</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8639043062)

**Comment by faddat** [comment]
so few commits!?

this is exciting indeed :D

File: x/eventmanager/keeper/events.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		if e.EventStatus == types.EventStatusPending {
+			res, err := e.CanExecute(ctx, &k)
+			if err != nil {
+				k.Logger(ctx).Error("unable to determine if event can execute callback", "module", e.Module, "id", e.Identifier, "callback", e.Callback, "error", err)
+			}
+			if res {
+				k.Logger(ctx).Info(fmt.Sprintf("triggered event callback %s for event %s (%s)", e.Callback, e.Identifier, e.ChainId))
+				err := k.Call(ctx, e.Module, e.Callback, e.Payload)
+				if err != nil {
+					k.Logger(ctx).Error("unable to execute callback", "module", e.Module, "id", e.Identifier, "callback", e.Callback, "error", err)
```
Comment (joe-bowman) [inline x/eventmanager/keeper/events.go:99]: We should buddle the error up here.

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>6da0983</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/CLbDUCJPMuBPTzkTWBhhhiJ36R1C'>https://vercel.com/quicksilver/quicksilver/CLbDUCJPMuBPTzkTWBhhhiJ36R1C</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8789100884)

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>e187605</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/DedZH9a6W7K1vivHpqJSN8P2dviP'>https://vercel.com/quicksilver/quicksilver/DedZH9a6W7K1vivHpqJSN8P2dviP</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8824104991)

**Comment by joe-bowman** [comment]
This pull request has been deployed to Vercel.

<table>
<tr>
<td><strong>Latest commit:</strong></td>
<td><code>86d8f46</code></td>
</tr>
<tr>
<td><strong>✅ Preview:</strong></td>
<td><a href='https://pr-1397.app-dev.quicksilver.zone'>https://pr-1397.app-dev.quicksilver.zone</a></td>
</tr>
<tr>
<td><strong>🔍 Inspect:</strong></td>
<td><a href='https://vercel.com/quicksilver/quicksilver/9pEwrhDrw2FqKHWY571iyrhxLToE'>https://vercel.com/quicksilver/quicksilver/9pEwrhDrw2FqKHWY571iyrhxLToE</a></td>
</tr>
</table>

[View Workflow Logs](https://github.com/quicksilver-zone/quicksilver/actions/runs/8846529126)

**Comment by joe-bowman** [comment]
This pull request has been automatically marked as stale because it has not had any recent activity. It will be closed if no further activity occurs. Thank you!

**Pull request closed without merging**